primitives/src/merkle_tree/internal.rs

```diff
@@ -576,7 +665,107 @@ where
                             height - 1,
                             &cur_pos,
                             traversal_path,
-                            tight_frontier && frontier == traversal_path[height - 1],
+                            at_frontier && frontier == traversal_path[height - 1],
+                            data,
+                        )?;
+                        cnt += increment;
+                        cur_pos += I::from(increment);
+                        frontier += 1;
+                    }
+                    *self = MerkleNode::Branch {
+                        value: digest_branch::<E, H, I, T>(&children),
+                        children,
+                    };
+                    Ok(cnt)
+                }
+            },
+            MerkleNode::Leaf { .. } => Err(PrimitivesError::ParameterError(
+                "Incompatible merkle tree: index already occupied".to_string(),
+            )),
+            MerkleNode::ForgettenSubtree { .. } => Err(PrimitivesError::ParameterError(
+                "Given part of merkle tree is not in memory".to_string(),
+            )),
+        }
+    }
+
+    pub(crate) fn extend_and_forget_internal<H, Arity>(
```

Comment:
Maybe add some comments explaining the API's functionality and the usage of the API arguments.